### ISSUE TITLE
`EarlyHintsRequestTest#test_when_early_hints_is_set_in_the_env_link_headers_are_sent` fails with since https://github.com/rack/rack/pull/1831

### ISSUE THREAD

**Issue by yahonda** [rails/rails#51961]
### Steps to reproduce

Managed to reproduce CI failure at https://buildkite.com/rails/rails-nightly/builds/598#018fcb4e-7d02-466b-852d-b109a35c38f9/1157-1212

```ruby
git clone https://github.com/rails/rails
cd rails/actionpack
export RACK="head"
bundle update rack
bin/test test/dispatch/request_test.rb:1415
```

### Expected behavior
It should pass.

### Actual behavior
It gets this error.

```ruby
$ bin/test test/dispatch/request_test.rb:1415
Running 116 tests in parallel using 4 processes
Run options: --seed 45076

# Running:

/home/yahonda/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/bundler/gems/rack-c8966d083db4/lib/rack/lint.rb:15: warning: URI.regexp is obsolete
/home/yahonda/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/bundler/gems/rack-c8966d083db4/lib/rack/lint.rb:15: warning: URI.regexp is obsolete
E

Error:
EarlyHintsRequestTest#test_when_early_hints_is_set_in_the_env_link_headers_are_sent:
Rack::Lint::LintError: uppercase character in header name: Link
    /home/yahonda/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/bundler/gems/rack-c8966d083db4/lib/rack/lint.rb:704:in `block in check_headers'
    /home/yahonda/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/bundler/gems/rack-c8966d083db4/lib/rack/lint.rb:688:in `each'
    /home/yahonda/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/bundler/gems/rack-c8966d083db4/lib/rack/lint.rb:688:in `check_headers'
    /home/yahonda/.rbenv/versions/3.3.2/lib/ruby/gems/3.3.0/bundler/gems/rack-c8966d083db4/lib/rack/lint.rb:656:in `block in check_early_hints'
    lib/action_dispatch/http/request.rb:242:in `send_early_hints'
    test/dispatch/request_test.rb:1415:in `block in <class:EarlyHintsRequestTest>'


bin/test test/dispatch/request_test.rb:1414



Finished in 0.640996s, 1.5601 runs/s, 0.0000 assertions/s.
1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
$
```

### System configuration
**Rails version**: main branch

**Ruby version**: ruby 3.3.2 (2024-05-30 revision e5a195edf6) [x86_64-linux]

**Comment by yahonda** [comment]
git bisect says this behavior change has been introduced since https://github.com/rack/rack/pull/1831 https://github.com/rack/rack/commit/1fbcf54289b9fae2765c1167376e837fe38b543c

**Comment by zzak** [comment]
This is also this deprecation warning but probably a separate issue:

```
/usr/local/bundle/bundler/gems/rack-c8966d083db4/lib/rack/utils.rb:567: warning: Status code :unprocessable_entity is deprecated and will be removed in a future version of Rack. Please use :unprocessable_content instead.
```
